File: day-02/js/main.ts
const DEBUG = false;
const debug = DEBUG
	? (...args: unknown[]) => {
		console.log("DEBUG:", ...args);
	}
	: () => {};

const input = Deno.readTextFileSync(Deno.args[0]);
const reports = input.trim().split("\n").map((report) =>
	report.split(" ").map((level) => Number(level))
);

debug("Loaded", reports.length, "reports");

const safeReports = reports.reduce(
	(safeReports, report) => safeReports + Number(isReportSafe(report, false)),
	0,
);
const safeReportsSkipped = reports.reduce(
	(safeReports, report) => safeReports + Number(isReportSafe(report, true)),
	0,
);

console.log("Total number of safe reports is:", safeReports);
console.log(
	"Total number of safe reports with 1 value skipped is:",
	safeReportsSkipped,
);

function isReportSafe(report: number[], canSkip: boolean): boolean {
	debug("Report is:", report);

	if (isSequenceSafe(report)) {
		return true;
	}

	if (!canSkip) {
		return false;
	}

	for (let i = 0; i < report.length; i++) {
		const newReport = report.slice(0, i).concat(report.slice(i + 1));
		if (isSequenceSafe(newReport)) {
			return true;
		}
	}

	return false;
}

function isSequenceSafe(report: number[]): boolean {
	let isIncreasing = null;

	for (let i = 1; i < report.length; i++) {
		const delta = report[i] - report[i - 1];

		if (Math.abs(delta) > 3 || delta === 0) {
			return false;
		}

		if (isIncreasing === null) {
			isIncreasing = delta > 0;
		} else if (isIncreasing && delta < 0) {
			return false;
		} else if (!isIncreasing && delta > 0) {
			return false;
		}
	}

	return true;
}

File: day-02/rust/src/main.rs
use std::{env, fs};

use itertools::Itertools;

fn main() {
    let reports = fs::read_to_string(
        env::args()
            .nth(1)
            .expect("input.txt file path not provided"),
    )
    .expect("couldn't read input.txt")
    .lines()
    .map(|line| {
        line.split(' ')
            .map(|val| val.parse().unwrap())
            .collect::<Vec<u32>>()
    })
    .collect::<Vec<_>>();

    let safe_reports = reports
        .iter()
        .map(|report| u32::from(is_report_safe(report, false)))
        .sum::<u32>();

    let safe_reports_skipped = reports
        .iter()
        .map(|report| u32::from(is_report_safe(report, true)))
        .sum::<u32>();

    println!("Total number of safe reports is: {safe_reports}");
    println!("Total number of safe reports with 1 value skipped is: {safe_reports_skipped}");
}

fn is_report_safe(report: &[u32], can_skip: bool) -> bool {
    if is_sequence_safe(report.iter().copied()) {
        return true;
    }

    if !can_skip {
        return false;
    }

    for i in 0..report.len() {
        let new_report = report[0..i].iter().chain(report[i + 1..].iter()).copied();
        if is_sequence_safe(new_report) {
            return true;
        }
    }

    false
}

fn is_sequence_safe(sequence: impl Iterator<Item = u32>) -> bool {
    let mut is_increasing = None;

    for (prev, next) in sequence.tuple_windows() {
        let delta = next as i32 - prev as i32;

        if delta.abs() > 3 || delta == 0 {
            return false;
        }

        match is_increasing {
            None => is_increasing = Some(delta > 0),
            Some(is_increasing) if is_increasing && delta < 0 => return false,
            Some(is_increasing) if !is_increasing && delta > 0 => return false,
            _ => (),
        }
    }

    true
}
